processor-api/src/main/java/com/synopsys/integration/alert/processor/api/summarize/DefaultProjectMessageSummarizer.java

```diff
@@ -0,0 +1,160 @@
+            case COMPONENT_UPDATE:
+            case COMPONENT_CONCERN:
+                return constructComponentStatusAndDescription(providerName, projectName, messageReason, projectMessage);
+            default:
+                return Pair.of(
+                    String.format("%s had notifications in BlackDuck", projectMessage.getProject().getValue()),
+                    "Alert received an unrecognized notification from BlackDuck."
+                );
+        }
+    }
+
+    private Pair<String, String> constructProjectVersionStatusSummaryAndDescription(String providerName, String projectName, ProjectMessage projectMessage) {
+        String operationString = projectMessage.getOperation()
+                                     .map(this::convertToParticiple)
+                                     .orElse("updated");
+
+        Optional<String> optionalProjectVersionName = projectMessage.getProjectVersion().map(LinkableItem::getValue);
+        if (optionalProjectVersionName.isPresent()) {
+            String projectVersionName = optionalProjectVersionName.get();
+            return Pair.of(
+                String.format("[%s] %s > %s %s", providerName, projectName, projectVersionName, operationString),
+                String.format("A project-version '%s' was %s was in the BlackDuck project '%s'.", projectVersionName, operationString, projectName)
+            );
+        } else {
+            return Pair.of(
+                String.format("[%s] %s %s", providerName, projectName, operationString),
+                String.format("The project '%s' was %s in BlackDuck.", projectName, operationString)
+            );
+        }
+    }
+
+    private Pair<String, String> constructComponentStatusAndDescription(String providerName, String projectName, MessageReason messageReason, ProjectMessage projectMessage) {
+        String projectVersionName = projectMessage.getProjectVersion()
+                                        .map(LinkableItem::getValue)
+                                        .orElse("Unknown Version");
+        String operationString = MessageReason.COMPONENT_CONCERN.equals(messageReason) ? "problems" : "updates";
+        return Pair.of(
+            String.format("[%s] %s > %s component %s", providerName, projectName, projectVersionName, operationString),
+            String.format("The project-version '%s > %s' had component %s in BlackDuck", projectName, projectVersionName, operationString)
+        );
+    }
+
+    private List<LinkableItem> constructMessageDetails(ProjectMessage projectMessage) {
+        List<LinkableItem> details = new LinkedList<>();
+        details.add(projectMessage.getProvider());
+        details.add(projectMessage.getProject());
+        projectMessage.getProjectVersion().ifPresent(details::add);
+
+        Map<Triple<ComponentConcernType, ItemOperation, ComponentConcernSeverity>, Integer> concernCounts = new LinkedHashMap<>();
```

Comment:
I think we should try avoiding Pair and Triple. When we make a class for these fields, we give them more of a purpose and can add functionality within the class to help with processing

Reply:
You and Jake make good points. I was concerned about having data classes that don't get exposed as part of an API, but I think it would help readability.

Reply:
Is this the class for a private inner class?

Reply:
Yes, but I'm also concerned about bloating the code.